akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -0,0 +1,390 @@
+      override def onPush(): Unit = {
+        onNext(grab(in))
+        if (downstreamCompleted) cancel(in)
+        else if (downstreamDemand > 0) pull(in)
+      }
+
+      override def onUpstreamFinish(): Unit = complete()
+
+      override def onUpstreamFailure(cause: Throwable): Unit = fail(cause)
+    })
+
+    def subscribePending(): Unit =
+      exposedPublisher.takePendingSubscribers() foreach { sub ⇒
+        if (subscriber eq null) {
+          subscriber = sub
+          tryOnSubscribe(subscriber, new BoundarySubscription(actor, id))
+        } else
+          rejectAdditionalSubscriber(subscriber, s"${Logging.simpleName(this)}")
+      }
+
+    def exposedPublisher(publisher: ActorPublisher[Any]): Unit = {
+      upstreamFailed match {
+        case _: Some[_] ⇒
+          publisher.shutdown(upstreamFailed)
+        case _ ⇒
+          exposedPublisher = publisher
+      }
+    }
+
+    def requestMore(elements: Long): Unit = {
+      if (elements < 1) {
+        cancel(in)
+        fail(ReactiveStreamsCompliance.numberOfElementsInRequestMustBePositiveException)
+      } else {
+        downstreamDemand += elements
+        if (downstreamDemand < 0)
+          downstreamDemand = Long.MaxValue // Long overflow, Reactive Streams Spec 3:17: effectively unbounded
+        if (!hasBeenPulled(in)) pull(in)
+      }
+    }
+
+    def cancel(): Unit = {
+      downstreamCompleted = true
+      subscriber = null
+      exposedPublisher.shutdown(Some(new ActorPublisher.NormalShutdownException))
+      cancel(in)
+    }
+
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] class ActorGraphInterpreter(assembly: GraphAssembly, shape: Shape, settings: ActorMaterializerSettings) extends Actor {
+  import ActorGraphInterpreter._
+
+  val interpreter = new GraphInterpreter(assembly, (logic, event, handler) ⇒ self ! AsyncInput(logic, event, handler))
+  val inputs = Array.tabulate(shape.inlets.size)(new BatchingActorInputBoundary(settings.maxInputBufferSize, _))
+  val outputs = Array.tabulate(shape.outlets.size)(new ActorOutputBoundary(self, _))
+  // Limits the number of events processed by the interpreter before scheduling a self-message for fairness with other
+  // actors.
+  // TODO: Better heuristic here
+  val eventLimit = settings.maxInputBufferSize * assembly.stages.length * 4 // Roughly 4 events per element transfer
+  // Limits the number of events processed by the interpreter on an abort event.
+  // TODO: Better heuristic here
+  val abortLimit = eventLimit * 2
+  var resumeScheduled = false
+
+  override def preStart(): Unit = {
+    var i = 0
+    while (i < inputs.length) {
+      interpreter.attachUpstreamBoundary(i, inputs(i))
+      i += 1
+    }
+    val offset = assembly.connectionCount - outputs.length
+    i = 0
+    while (i < outputs.length) {
+      interpreter.attachDownstreamBoundary(i + offset, outputs(i))
+      i += 1
+    }
+    interpreter.init()
+  }
+
+  override def receive: Receive = {
```

Comment:
Makes sense, and OnNext should be the first, as RequestMore is batched. I will move and comment that the order is important.